kong/plugins/prometheus/wasmx.lua

```diff
@@ -0,0 +1,234 @@
+local buffer = require "string.buffer"
+local wasm = require "kong.runloop.wasm"
+local wasmx_shm
+
+
+local pcall = pcall
+local str_sub = string.sub
+local table_insert = table.insert
+local table_sort = table.sort
+local buf_new = buffer.new
+local ngx_say = ngx.say
+local ngx_re_match = ngx.re.match
+
+
+local _M = {}
+
+
+local FLUSH_EVERY = 100
+local GET_METRIC_OPTS = { prefix = false }
+
+
+local metrics_data_buf = buf_new()
+local labels_serialization_buf = buf_new()
+local sum_lines_buf = buf_new()
+local count_lines_buf = buf_new()
+
+
+local function sorted_iter(ctx, i)
+  i = i + 1
+
+  local v = ctx.t[ctx.sorted_keys[i]]
+
+  if v ~= nil then
+    return i, v
+  end
+end
+
+
+local function sorted_pairs(t)
+  local sorted_keys = {}
+
+  for k, _ in pairs(t) do
+    table_insert(sorted_keys, k)
+  end
+
+  table_sort(sorted_keys)
+
+  return sorted_iter, { t = t, sorted_keys = sorted_keys }, 0
+end
+
+--
+-- Convert a pw_key into a pair of metric name and labels
+--
+-- pw_key follows the form `pw:<filter_name>:<metric_name>`
+-- `<metric_name>` might contain labels, e.g. a_metric_label1="v1";
+-- if it does, the position of the first label corresponds to the end of the
+-- metric name and is used to discard labels from <metric_name>.
+local function parse_pw_key(pw_key)
+  local m_name = pw_key
+  local m_labels = {}
+  local m_1st_label_pos = #pw_key
+
+  local matches = ngx_re_match(pw_key, [[pw:([\w\.]+):]])
+  local f_name = matches[1]
+  local f_meta = wasm.filter_meta[f_name] or {}
+  local l_patterns = f_meta.metrics and f_meta.metrics.label_patterns or {}
+
+  local match_ctx = {}
+
+  for _, pair in ipairs(l_patterns) do
+    matches = ngx_re_match(pw_key, pair.pattern, "", match_ctx)
+
+    if matches then
+      local l_pos, value = match_ctx.pos - #matches[1], matches[2]
+
+      table_insert(m_labels, { pair.label, value })
+
+      m_1st_label_pos = (l_pos < m_1st_label_pos) and l_pos or m_1st_label_pos
+    end
+  end
+
+  if m_1st_label_pos ~= #pw_key then
+    -- discarding labels from m_name
+    m_name = str_sub(pw_key, 1, m_1st_label_pos - 1)
+  end
+
+  return m_name, m_labels
+end
+
+
+--
+-- Parse potential labels stored in the metric key
+--
+-- If no labels are present, key is simply the metric name.
+local function parse_key(key)
+  local name = key
+  local labels
+
+  if #key > 3 and key:sub(0, 3) == "pw:" then
```

Comment:
nit: code works fine with the 0 index, but 1 is more correct for Lua semantics

```suggestion
  if #key > 3 and key:sub(1, 3) == "pw:" then
```